### PR TITLE
[google_maps_flutter_web] Initial support for custom overlays

### PR DESCRIPTION
**Pull request by elitree**
This is a resubmission of https://github.com/flutter/plugins/pull/6982 from the now archived flutter plugins repo. I'm submitting the changes from the original author, @AsturaPhoenix. The original description is below.

--------

Saves tile bytes to blobs and uses img elements to decode and render. Does not implement opacity, perform caching, or serve placeholder images.

**Issue:** Fixes https://github.com/flutter/flutter/issues/98596

**Known issues:**

- https://github.com/flutter/flutter/issues/116132
- https://github.com/AsturaPhoenix/trip_planner_aquamarine/issues/22

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart
Comment (elitree) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart:46]: [Original comment from @AsturaPhoenix](https://github.com/flutter/plugins/pull/6982#discussion_r1072927569):

We might want to put this in a `finally`. On the other hand, not revoking the blob URL in the case of a decoding failure may actually make debugging easier.

File: packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.dart
Comment (elitree) [inline packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/google_maps_controller_test.dart:357]: [Original comment by @AsturaPhoenix](https://github.com/flutter/plugins/pull/6982/files/c24d187970545ee18a52c00e8538c823a83fcf65#r1102025301):

Leaving a comment here because it wasn't obvious to me the second time I looked at it:

This test was removed because it became trivial once using Mockito verification; it essentially tests to make sure that `Marker.infoWindow` defaults to `InfoWindow.noText`. On the other hand, the argument could be made to keep it anyway (e.g. to construct the verification with an explicit `infoWindow: InfoWindow.noText`); let me know if you'd prefer that.

**Comment by elitree** [comment]
Resolved a changelog conflict and rebased on main.

**Comment by AsturaPhoenix** [comment]
FYI I just moved over my local repo as well, so that's one less hurdle if we do end up needing to collaborate. Thanks for taking over @elitree!

**Comment by stuartmorgan** [comment]
Update from triage: waiting for @ditman to have review bandwidth for maps PRs.

**Comment by elitree** [comment]
Rebased on main to resolve conflicts.

**Comment by stuartmorgan** [comment]
@ditman Is this still in your review queue?

**Comment by ditman** [comment]
Absolutely

File: packages/google_maps_flutter/google_maps_flutter_web/pubspec.yaml
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/pubspec.yaml:20]: I don't think we need this just to insert in the z-index sorted list (I'll take a look!)

**Review by ditman — COMMENTED** [review]
I want to tinker with this PR, is there any `example` that uses the tile overlays feature so I can see it working? Maybe the google_maps_flutter example app?

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (elitree) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:67]: I should have added a comment for this. With the 'auto' setting, using the scroll wheel to zoom requires holding a modifier key and the user gets an instruction like the following:
Screenshot 2023-05-23 at 11 13 00 AM

We changed the setting to 'greedy' so that the map could be zoomed with the scroll wheel without using the modifier key.

Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart:67]: @elitree please, revert this for now! There's more proper web gesture handling incoming here:

https://github.com/flutter/packages/pull/3258/

**Comment by ditman** [comment]
PS: There seems some tests breaking, it seems that the Mocks need to be regenerated (and committed):

https://cirrus-ci.com/task/5045655211606016?logs=drive#L55-L73

(Mocks can be (re)generated with: `dart run build_runner build --delete-conflicting-outputs` inside the `example` directory IIRC; there should be a `build.yaml` file with a little bit of config for the build runner)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart
Comment (elitree) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart]: I rebased this morning, and reverted this change.

**Comment by elitree** [comment]
> PS: There seems some tests breaking, it seems that the Mocks need to be regenerated (and committed):
> 
> https://cirrus-ci.com/task/5045655211606016?logs=drive#L55-L73
> 
> (Mocks can be (re)generated with: `dart run build_runner build --delete-conflicting-outputs` inside the `example` directory IIRC; there should be a `build.yaml` file with a little bit of config for the build runner)

I updated the mocks today as well, thanks for the reminder.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart]: I changed this particular bit, instead of `await img.decode`, I moved the revokeObjectUrl to the `load` event of `img`, so it happens... eventually.

**Comment by ditman** [comment]
I had some code ready to remove `package:collection`, but I just noticed that it's already used in the plugin [here](https://github.com/flutter/packages/blob/main/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/polygon.dart#L5), so I guess it can stay 😭

File: packages/google_maps_flutter/google_maps_flutter_web/pubspec.yaml
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/pubspec.yaml:20]: (This is already used by the platform_interface, so it can stay)

File: packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlay_test.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlay_test.dart:64]: We used to expect that `img.hidden` was `true` before load, and `false` afterwards. Is this really required? In my testing I didn't see any difference and removed the attribute. @AsturaPhoenix / @elitree do you remember why was the attribute added?

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart
Comment (ditman) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart:11]: I am very curious about this `logicalTileSize`. From the API it seems that both a TileOverlay and each Tile can specify their size, so this implementation can do away with this `logicalTileSize` const.

Why is this fixed to 256, and do we need it?

(If this is 256, how come the example is [creating 100x100 tiles](https://github.com/flutter/packages/blob/main/packages/google_maps_flutter/google_maps_flutter/example/lib/tile_overlay.dart#L119-L120)? They look blurry on the web as they're scaled up to fit!)

(Also: can tiles be non-square??)

File: packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlay_test.dart
Comment (AsturaPhoenix) [reply packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlay_test.dart:64]: I think it was to prevent the image decoding from blocking the main thread, but I can't find a version without hidden + decode so I'm not sure I ever ran a comparison. In any case we still have the bottleneck of populating the image data on the Dart/Flutter side, which has to happen on the main thread on web today.

I do think there's a difference in the case of bad image data though; if `Tile.data` is not a valid image, I think adding the `img` straight and setting the `src` will eventually make it display a browser-dependent broken image whereas hiding and decoding wouldn't, but I don't think that was the primary motivation.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart:11]: @AsturaPhoenix yes, I used the physical size being `logicalSize * dpi` to generate high DPI tiles too. Let's go with consistency for now, having a fixed logical size.

(PS: In iOS they also mention 256x256 tiles or 512x512 for "retina" displays: https://developers.google.com/maps/documentation/ios-sdk/tiles#high_dpi_tiles_for_retina_devices, I'm not sure if the iOS SDK supports sizes any other than those two (like a 3x tile))

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart
Comment (AsturaPhoenix) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart:32]: Oh, cool.

Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlay.dart:32]: I brought back `hidden` as well, I'm not sure if it's because I was rendering different tiles, but it did seem faster to scroll the map around with it... No scientific perf measure there, just used my "feelings" haha

**Comment by ditman** [comment]
OK, I changed the Overlays controller a little bit so it looks and behaves more similarly to other geometry controllers in the plugin.

Please @elitree and @AsturaPhoenix give my latest changes a shot to see if the plugin still behaves the way you need. In the meantime I'll find a 2nd reviewer for t his to land!

Again, thanks for the PR! Very exciting new feature incoming!

**Comment by ditman** [comment]
Adding @stuartmorgan as 2nd reviewer.

**Review by ditman — APPROVED** [review]
LGTM after doing some tweaking, but possibly missed stuff.

**Comment by elitree** [comment]
This looks good to me, thanks! 

I think @AsturaPhoenix wound up adding some other performance changes elsewhere in other branches, but this is an excellent changeset for initial support IMO.

**Comment by ditman** [comment]
When testing with the `example` app, the real performance hit happened here:

* https://github.com/flutter/packages/blob/442d9db34717454e99cd27b188724b858d72227e/packages/google_maps_flutter/google_maps_flutter/example/lib/tile_overlay.dart#L146-L150

Not so much on this side of the implementation :)

**Comment by AsturaPhoenix** [comment]
Yeah, unfortunately the other performance changes probably won't pass muster since they involve changing the API to pass canvases as tiles instead of byte arrays. (There are also unrelated optimizations around serialization of polylines and friends, but unfortunately eventually I ran into an expensive serialization we didn't seem able to avoid within GMS core, so that was the end of that.)

This LGTM too; I have one nit comment that I'll add, but it's probably not of consequence.

**Comment by AsturaPhoenix** [comment]
> When testing with the `example` app, the real performance hit happened here:
> 
> * https://github.com/flutter/packages/blob/442d9db34717454e99cd27b188724b858d72227e/packages/google_maps_flutter/google_maps_flutter/example/lib/tile_overlay.dart#L146-L150
> 
> Not so much on this side of the implementation :)
> 
> (`toImage` and `toByteData` being very expensive)

Yes, it's quite painful. We can get a few ms back by using bmps instead of pngs, but we don't get jank-free until we pass canvases afaict. (flutter/flutter#116132)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart:70]: This little bit of code concerns me a little bit as well, but we'll see in the real world :) I simplified the original if-else a little bit and ended up with this, and I might have over-simplified. I'll tag you whenever the first bug comes about this so you can tell me: "I TOLD YOU SO!" :p

**Comment by ditman** [comment]
@AsturaPhoenix in this case I see two optimizations:

* Do not generate the Uint8List in the main thread on the web (that means something like this: https://developer.chrome.com/blog/offscreen-canvas/)
* Make `Tile.data` an `Object?`, so each platform can pass what's most efficient to its implementation (in mobile it might be a byte array, but as you said on the web it might be a straight-up canvas element)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart
Comment (AsturaPhoenix) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart:70]: Oh actually you know what, rereading this more carefully I think it does warrant another pass. For example, there actually seems to be a bug (that I almost doubt anyone would hit) where if you change the Z index of a layer that's not visible, it becomes visible. (Purely based on my reading of the code; I haven't run a test, so I might be wrong.)

And, since this is in a loop, the impact of remove/insert vs. set for the stable z-index case, although probably still a premature optimization, could be compounded if it were ever actually an issue.

Comment (AsturaPhoenix) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart:70]: Basically the fix if we want to do it this way is just to remove the `moved` condition entirely, but you know what I'd do if it were up to me.

**Review by AsturaPhoenix — COMMENTED** [review]
Okay, I don't have permission to push to the remote, but here's the patch (also adding some additional assertions in related tests to try to catch similar regressions):

```diff
diff --git a/packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlays_test.dart b/packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlays_test.dart
index 4b95d8bbe..8b6b34694 100644
--- a/packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlays_test.dart
+++ b/packages/google_maps_flutter/google_maps_flutter_web/example/integration_test/overlays_test.dart
@@ -68,6 +68,9 @@ void main() {
         tileProviders[1].getTile(any, any, any),
         tileProviders[2].getTile(any, any, any),
       ]);
+      verifyNoMoreInteractions(tileProviders[0]);
+      verifyNoMoreInteractions(tileProviders[1]);
+      verifyNoMoreInteractions(tileProviders[2]);
     });

     testWidgets('changeTileOverlays', (WidgetTester tester) async {
@@ -87,6 +90,8 @@ void main() {
         tileProviders[1].getTile(any, any, any),
       ]);
       verifyZeroInteractions(tileProviders[0]);
+      verifyNoMoreInteractions(tileProviders[1]);
+      verifyNoMoreInteractions(tileProviders[2]);

       // Re-enable overlay 0.
       controller.changeTileOverlays(
@@ -99,6 +104,22 @@ void main() {
         tileProviders[0].getTile(any, any, any),
         tileProviders[1].getTile(any, any, any),
       ]);
+      verifyNoMoreInteractions(tileProviders[0]);
+      verifyNoMoreInteractions(tileProviders[1]);
+      verifyNoMoreInteractions(tileProviders[2]);
+    });
+
+    testWidgets(
+        'updating the z index of a hidden layer does not make it visible',
+        (WidgetTester tester) async {
+      controller.addTileOverlays(<TileOverlay>{...tileOverlays});
+
+      controller.changeTileOverlays(<TileOverlay>{
+        tileOverlays[0].copyWith(zIndexParam: -1, visibleParam: false),
+      });
+
+      probeTiles();
+      verifyZeroInteractions(tileProviders[0]);
     });

     testWidgets('removeTileOverlays', (WidgetTester tester) async {
diff --git a/packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart b/packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart
index 464fb319a..aa6c19173 100644
--- a/packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart
+++ b/packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart
@@ -63,14 +63,13 @@ class TileOverlaysController extends GeometryController {

     final bool wasVisible = controller.tileOverlay.visible;
     final bool isVisible = tileOverlay.visible;
-    final bool moved = tileOverlay.zIndex != controller.tileOverlay.zIndex;

     controller.update(tileOverlay);

-    if (wasVisible || moved) {
+    if (wasVisible) {
       _remove(controller);
     }
-    if (isVisible || moved) {
+    if (isVisible) {
       _insertZSorted(controller);
     }
   }
```

**Comment by ditman** [comment]
@AsturaPhoenix, @elitree thanks for the fix over the weekend!

**Pull request merged**
